packages/date-picker/src/vaadin-date-picker-overlay-content.js

```diff
@@ -936,7 +969,11 @@ class DatePickerOverlayContent extends ControllerMixin(ThemableMixin(DirMixin(Po
     // Wait for `vaadin-month-calendar` elements to be rendered
     if (!this.calendars.length) {
       await new Promise((resolve) => {
-        setTimeout(resolve);
+        afterNextRender(this, () => {
+          // Force dom-repeat elements to render
+          flush();
+          resolve();
+        });
```

Comment:
Replaced `setTimeout` with a more reliable logic to ensure date cells are actually rendered.
This is a proper fix for the issue that I previously attempted to fix in #4296.